osu.Game/Rulesets/Mods/IApplicableFailOverride.cs

```diff
@@ -1,22 +1,43 @@
 ﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using osu.Game.Rulesets.Judgements;
+
 namespace osu.Game.Rulesets.Mods
 {
     /// <summary>
-    /// Represents a mod which can override (and block) a fail.
+    /// Represents a mod which can override failure, by either hard-blocking it, or forcing it immediately.
     /// </summary>
     public interface IApplicableFailOverride : IApplicableMod
     {
         /// <summary>
-        /// Whether we should allow failing at the current point in time.
+        /// Whether we want to restart on fail.
         /// </summary>
-        /// <returns>Whether the fail should be allowed to proceed. Return false to block.</returns>
-        bool PerformFail();
+        bool RestartOnFail { get; }
 
         /// <summary>
-        /// Whether we want to restart on fail. Only used if <see cref="PerformFail"/> returns true.
+        /// Check the current failure allowance for this mod.
         /// </summary>
-        bool RestartOnFail { get; }
+        /// <param name="result">The judgement result which should be considered. Importantly, will be <c>null</c> if a failure has already being triggered.</param>
```

Comment:
The `null` condition really destroys my brain :/

Reply:
I agree, and I'd like to be able to avoid it, but I couldn't find a good way to get the responsible judgement for already triggered cases.

Reply:
I tried adding a `bool` but I'm honestly not sure it helps much. Updating a lot of the usages from `result == null` to `failTriggered` requires a secondary null check regardless (so unless there's some fancy cross-parameter `NotNullWhenParamIsTrue` attribute I'm missing...)

```diff
diff --git a/osu.Game.Rulesets.Mania/Mods/ManiaModPerfect.cs b/osu.Game.Rulesets.Mania/Mods/ManiaModPerfect.cs
index 24dcf67bca..450f2c07f2 100644
--- a/osu.Game.Rulesets.Mania/Mods/ManiaModPerfect.cs
+++ b/osu.Game.Rulesets.Mania/Mods/ManiaModPerfect.cs
@@ -9,7 +9,7 @@ namespace osu.Game.Rulesets.Mania.Mods
 {
     public class ManiaModPerfect : ModPerfect
     {
-        public override FailState CheckFail(JudgementResult? result)
+        public override FailState CheckFail(bool failTriggered, JudgementResult? result)
         {
             if (result == null)
                 return FailState.Allow;
diff --git a/osu.Game.Rulesets.Osu/Mods/OsuModTargetPractice.cs b/osu.Game.Rulesets.Osu/Mods/OsuModTargetPractice.cs
index 5afaec4868..6f7fb4606a 100644
--- a/osu.Game.Rulesets.Osu/Mods/OsuModTargetPractice.cs
+++ b/osu.Game.Rulesets.Osu/Mods/OsuModTargetPractice.cs
@@ -102,7 +102,7 @@ public class OsuModTargetPractice : ModWithVisibilityAdjustment, IApplicableToDr
 
         public bool RestartOnFail => false;
 
-        public FailState CheckFail(JudgementResult? result)
+        public FailState CheckFail(bool failTriggered, JudgementResult? result)
         {
             if (result == null)
                 return FailState.Allow;
diff --git a/osu.Game.Tests/Gameplay/TestSceneDrainingHealthProcessor.cs b/osu.Game.Tests/Gameplay/TestSceneDrainingHealthProcessor.cs
index 7b575183c8..299f1cc41d 100644
--- a/osu.Game.Tests/Gameplay/TestSceneDrainingHealthProcessor.cs
+++ b/osu.Game.Tests/Gameplay/TestSceneDrainingHealthProcessor.cs
@@ -440,7 +440,7 @@ public ModFailOnResult(HitResult type)
                 this.type = type;
             }
 
-            public override FailState CheckFail(JudgementResult result) => result?.Type == type ? FailState.Force : FailState.Allow;
+            public override FailState CheckFail(bool failTriggered, JudgementResult result) => result?.Type == type ? FailState.Force : FailState.Allow;
         }
     }
 }
diff --git a/osu.Game.Tests/Visual/Gameplay/TestSceneFailJudgement.cs b/osu.Game.Tests/Visual/Gameplay/TestSceneFailJudgement.cs
index 964a674108..e88130dfcc 100644
--- a/osu.Game.Tests/Visual/Gameplay/TestSceneFailJudgement.cs
+++ b/osu.Game.Tests/Visual/Gameplay/TestSceneFailJudgement.cs
@@ -61,7 +61,7 @@ private class ModFailOnFirstJudgement : ModFailCondition
                 public override double ScoreMultiplier => 1.0;
                 public override string Acronym => "";
 
-                public override FailState CheckFail(JudgementResult? result) => FailState.Force;
+                public override FailState CheckFail(bool failTriggered, JudgementResult? result) => FailState.Force;
             }
         }
     }
diff --git a/osu.Game/Rulesets/Mods/IApplicableFailOverride.cs b/osu.Game/Rulesets/Mods/IApplicableFailOverride.cs
index 7c47261a21..013788a368 100644
--- a/osu.Game/Rulesets/Mods/IApplicableFailOverride.cs
+++ b/osu.Game/Rulesets/Mods/IApplicableFailOverride.cs
@@ -18,9 +18,10 @@ public interface IApplicableFailOverride : IApplicableMod
         /// <summary>
         /// Check the current failure allowance for this mod.
         /// </summary>
+        /// <param name="failTriggered">Whether a failure has been triggered. In this state, only a <see cref="FailState.Block"/> will change the outcome.</param>
         /// <param name="result">The judgement result which should be considered. Importantly, will be <c>null</c> if a failure has already being triggered.</param>
         /// <returns>The current failure allowance (see <see cref="FailState"/>).</returns>
-        FailState CheckFail(JudgementResult? result);
+        FailState CheckFail(bool failTriggered, JudgementResult? result);
     }
 
     public enum FailState
diff --git a/osu.Game/Rulesets/Mods/ModAccuracyChallenge.cs b/osu.Game/Rulesets/Mods/ModAccuracyChallenge.cs
index 1176a452e0..f82f8ed344 100644
--- a/osu.Game/Rulesets/Mods/ModAccuracyChallenge.cs
+++ b/osu.Game/Rulesets/Mods/ModAccuracyChallenge.cs
@@ -74,7 +74,7 @@ public void ApplyToScoreProcessor(ScoreProcessor scoreProcessor)
 
         public ScoreRank AdjustRank(ScoreRank rank, double accuracy) => rank;
 
-        public override FailState CheckFail(JudgementResult? result) => FailState.Allow;
+        public override FailState CheckFail(bool failTriggered, JudgementResult? result) => FailState.Allow;
 
         public enum AccuracyMode
         {
diff --git a/osu.Game/Rulesets/Mods/ModCinema.cs b/osu.Game/Rulesets/Mods/ModCinema.cs
index 70a0138f8a..b45889ca27 100644
--- a/osu.Game/Rulesets/Mods/ModCinema.cs
+++ b/osu.Game/Rulesets/Mods/ModCinema.cs
@@ -51,6 +51,6 @@ public void ApplyToPlayer(Player player)
 
         public bool RestartOnFail => false;
 
-        public FailState CheckFail(JudgementResult? result) => FailState.Block;
+        public FailState CheckFail(bool failTriggered, JudgementResult? result) => FailState.Block;
     }
 }
diff --git a/osu.Game/Rulesets/Mods/ModEasyWithExtraLives.cs b/osu.Game/Rulesets/Mods/ModEasyWithExtraLives.cs
index 891d76e48a..6825d661cc 100644
--- a/osu.Game/Rulesets/Mods/ModEasyWithExtraLives.cs
+++ b/osu.Game/Rulesets/Mods/ModEasyWithExtraLives.cs
@@ -36,9 +36,9 @@ public override void ApplyToDifficulty(BeatmapDifficulty difficulty)
 
         public bool RestartOnFail => false;
 
-        public FailState CheckFail(JudgementResult? result)
+        public FailState CheckFail(bool failTriggered, JudgementResult? result)
         {
-            if (result != null)
+            if (!failTriggered)
                 return FailState.Block;
 
             if (retries == 0)
diff --git a/osu.Game/Rulesets/Mods/ModFailCondition.cs b/osu.Game/Rulesets/Mods/ModFailCondition.cs
index 9cb9d25be4..448a1bb9dd 100644
--- a/osu.Game/Rulesets/Mods/ModFailCondition.cs
+++ b/osu.Game/Rulesets/Mods/ModFailCondition.cs
@@ -31,6 +31,6 @@ public void ApplyToHealthProcessor(HealthProcessor healthProcessor)
         /// </summary>
         protected void TriggerFailure() => triggerFailureDelegate?.Invoke(this);
 
-        public abstract FailState CheckFail(JudgementResult? result);
+        public abstract FailState CheckFail(bool failTriggered, JudgementResult? result);
     }
 }
diff --git a/osu.Game/Rulesets/Mods/ModNoFail.cs b/osu.Game/Rulesets/Mods/ModNoFail.cs
index e64c2a078c..1f3b448246 100644
--- a/osu.Game/Rulesets/Mods/ModNoFail.cs
+++ b/osu.Game/Rulesets/Mods/ModNoFail.cs
@@ -30,7 +30,7 @@ public abstract class ModNoFail : Mod, IApplicableFailOverride, IApplicableToHUD
         /// <summary>
         /// We never fail, 'yo.
         /// </summary>
-        public FailState CheckFail(JudgementResult? result) => FailState.Block;
+        public FailState CheckFail(bool failTriggered, JudgementResult? result) => FailState.Block;
 
         public void ReadFromConfig(OsuConfigManager config)
         {
diff --git a/osu.Game/Rulesets/Mods/ModPerfect.cs b/osu.Game/Rulesets/Mods/ModPerfect.cs
index a4c6192f55..07be8ef79a 100644
--- a/osu.Game/Rulesets/Mods/ModPerfect.cs
+++ b/osu.Game/Rulesets/Mods/ModPerfect.cs
@@ -28,7 +28,7 @@ protected ModPerfect()
             Restart.Value = Restart.Default = true;
         }
 
-        public override FailState CheckFail(JudgementResult? result)
+        public override FailState CheckFail(bool failTriggered, JudgementResult? result)
         {
             if (result == null)
                 return FailState.Allow;
diff --git a/osu.Game/Rulesets/Mods/ModSuddenDeath.cs b/osu.Game/Rulesets/Mods/ModSuddenDeath.cs
index 49b3430f24..a462e2bbdb 100644
--- a/osu.Game/Rulesets/Mods/ModSuddenDeath.cs
+++ b/osu.Game/Rulesets/Mods/ModSuddenDeath.cs
@@ -23,7 +23,7 @@ public abstract class ModSuddenDeath : ModFailCondition
 
         public override Type[] IncompatibleMods => base.IncompatibleMods.Append(typeof(ModPerfect)).ToArray();
 
-        public override FailState CheckFail(JudgementResult? result)
+        public override FailState CheckFail(bool failTriggered, JudgementResult? result)
         {
             if (result == null)
                 return FailState.Allow;
diff --git a/osu.Game/Rulesets/Scoring/HealthProcessor.cs b/osu.Game/Rulesets/Scoring/HealthProcessor.cs
index 793256abb4..69b9e7f519 100644
--- a/osu.Game/Rulesets/Scoring/HealthProcessor.cs
+++ b/osu.Game/Rulesets/Scoring/HealthProcessor.cs
@@ -117,7 +117,7 @@ private bool meetsAnyFailCondition(JudgementResult result)
 
             foreach (var condition in OrderedFailOverrideMods)
             {
-                if (condition.CheckFail(result) == FailState.Force)
+                if (condition.CheckFail(false, result) == FailState.Force)
                 {
                     ModTriggeringFailure = condition as Mod;
                     return true;
diff --git a/osu.Game/Screens/Play/Player.cs b/osu.Game/Screens/Play/Player.cs
index 4d8e4d1383..1c29580bdb 100644
--- a/osu.Game/Screens/Play/Player.cs
+++ b/osu.Game/Screens/Play/Player.cs
@@ -152,7 +152,7 @@ public abstract partial class Player : ScreenWithBeatmapBackground, ISamplePlayb
         /// Whether failing should be allowed.
         /// By default, this checks whether all selected mods allow failing.
         /// </summary>
-        protected virtual bool CheckModsAllowFailure() => HealthProcessor.OrderedFailOverrideMods.All(m => m.CheckFail(null) != FailState.Block);
+        protected virtual bool CheckModsAllowFailure() => HealthProcessor.OrderedFailOverrideMods.All(m => m.CheckFail(true, null) != FailState.Block);
 
         public readonly PlayerConfiguration Configuration;
 

```

Reply:
To explain a bit further, my problem with this occurs when reading `ModEasyWithExtraLives`. It reads as "if any judgement has occurred, then we block fail", and, with my prior knowledge that failure can _only_ occur as a result of a judgement, the rest of the code in that method looks like an oxymoron.

It's only when you realise that _regardless of any of the mod code running_, `TriggerFail()` will still be invoked via the `CheckDefaultFailCondition()` path, and that later on there's a precondition inside `Player` calling the method a second time, that it all starts to make sense.  
It's very hidden and non-obvious logic, and I would expect a first time reader/user of this code to go along with my initial thought process.

Then we have to consider the second path that this can happen, which is if a mod triggers failure `ModAccuracyChallenge`, which is truly a path "without a judgement".

If my other comments in `HealthProcessor` sound reasonable, then maybe a simple solution is:

1) Invert the logic:
```
if (check_default_condition)
    TriggerFailure()

if (any_mod_forces_fail)
    TriggerFailure()
```
to:
```
if (force || allow && check_default_condition)
    TriggerFailure()
```

2) Replace the precondition inside `Player.onFail()`, so that it doesn't do a second lookup on `CheckFail()`.

3) Make `HealthProcessor.TriggerFailure()` private. Remove the ability for mods to _randomly_ trigger failure. We only allow failure on a judgement in principle, and the new state logic should be able to handle that. `ModAccuracyChallenge` could be implemented as forcing the fail provided the condition matches _during a judgement_.